dot/core/service.go

```diff
@@ -602,6 +603,142 @@ func (s *Service) createBlockResponse(msg *network.BlockRequestMessage) (*networ
 // chain by calling `core_execute_block`. Valid blocks are stored in the block
 // database to become part of the canonical chain.
 func (s *Service) ProcessBlockResponseMessage(msg network.Message) error {
+	log.Trace("[core] got BlockResponseMessage")
+
+	blockData := msg.(*network.BlockResponseMessage).BlockData
+
+	bestNum, err := s.blockState.BestBlockNumber()
+	if err != nil {
+		return err
+	}
+
+	for _, bd := range blockData {
+		if bd.Header.Exists() {
+			header, err := types.NewHeaderFromOptional(bd.Header)
+			if err != nil {
+				return err
+			}
+
+			// get block header; if exists, return
+			existingHeader, err := s.blockState.GetHeader(bd.Hash)
+			if err != nil && existingHeader == nil {
+				err = s.blockState.SetHeader(header)
+				if err != nil {
+					return err
+				}
+
+				log.Info("[core] saved block header", "hash", header.Hash(), "number", header.Number)
+
+				// TODO: handle consensus digest, if first in epoch
+				// err = s.handleConsensusDigest(header)
+				// if err != nil {
+				// 	return err
+				// }
+			}
+		}
+
+		if bd.Header.Exists() && bd.Body.Exists {
+			header, err := types.NewHeaderFromOptional(bd.Header)
+			if err != nil {
+				return err
+			}
+
+			body, err := types.NewBodyFromOptional(bd.Body)
+			if err != nil {
+				return err
+			}
+
+			block := &types.Block{
+				Header: header,
+				Body:   body,
+			}
+
+			// prepare block for sending to core_executeBlock,
+			//  core_executeBlock fails if Digest and Body data are sent
+			blockData := types.Block{
+				Header: &types.Header{
+					ParentHash:     header.ParentHash,
+					Number:         header.Number,
+					StateRoot:      header.StateRoot,
+					ExtrinsicsRoot: header.ExtrinsicsRoot,
+				},
+				Body: types.NewBody([]byte{}),
+			}
+
+			bdEnc, err := blockData.Encode()
+			if err != nil {
+				return err
+			}
+
+			res, err := s.executeBlock(bdEnc)
+			if err != nil {
+				log.Error("[core] failed to validate block", "err", err)
+				return err
+			}
+			// if executeBlock return a non-empty byte array something when wrong
+			if !reflect.DeepEqual(res, []byte{}) {
+				log.Error("[core] execute block call failed", "err", res)
+			}
```

Comment:
looks good, when you merge with development the logic for this function is moved to syncer.handleBlockResponse so you can move this there

Reply:
Any suggestion how I can get reference to runtime from within syncer.handleBlockResponse to call executeBlock?  Should I add reference to runtime in the syncer struct?  Would that be ugly because than syncer would require runtime?  Suggestions?

Reply:
I think adding runtime to the syncer for now is okay, it would be nice to have all the runtime functions in one place but we can figure that out later

Reply:
ok

Reply:
This was mention in #709 as well, moving it to syncer would be ok but all the runtime methods should be in one place. I will address in #709 or open a new issue if you do not address here.